drivers/serial/uart_bcm2711.c

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright (c) 2023 honglin leng <a909204013@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT brcm_bcm2711_aux_uart
+
+/**
+ * @brief BCM2711 Miniuart Serial Driver
+ *
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/arch/cpu.h>
+#include <stdbool.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/irq.h>
+
+#define BCM2711_MU_IO			0x00
+#define BCM2711_MU_IER			0x04
+#define BCM2711_MU_IIR			0x08
+#define BCM2711_MU_LCR			0x0c
+#define BCM2711_MU_MCR			0x10
+#define BCM2711_MU_LSR			0x14
+#define BCM2711_MU_MSR			0x18
+#define BCM2711_MU_SCRATCH		0x1c
+#define BCM2711_MU_CNTL			0x20
+#define BCM2711_MU_STAT			0x24
+#define BCM2711_MU_BAUD			0x28
+
+#define BCM2711_MU_IER_TX_INTERRUPT	BIT(1)
+#define BCM2711_MU_IER_RX_INTERRUPT	BIT(0)
+
+#define BCM2711_MU_IIR_RX_INTERRUPT	BIT(2)
+#define BCM2711_MU_IIR_TX_INTERRUPT	BIT(1)
+#define BCM2711_MU_IIR_FLUSH		0xc6
+
+#define BCM2711_MU_LCR_7BIT		0x02
+#define BCM2711_MU_LCR_8BIT		0x03
+
+#define BCM2711_MU_LSR_TX_IDLE		BIT(6)
+#define BCM2711_MU_LSR_TX_EMPTY		BIT(5)
+#define BCM2711_MU_LSR_RX_OVERRUN	BIT(1)
+#define BCM2711_MU_LSR_RX_READY		BIT(0)
+
+#define BCM2711_MU_CNTL_RX_ENABLE	BIT(0)
+#define BCM2711_MU_CNTL_TX_ENABLE	BIT(1)
+
+struct bcm2711_uart_config {
+	DEVICE_MMIO_ROM; /* Must be first */
+	uint32_t baud_rate;
+	uint32_t clocks;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	void (*irq_config_func)(const struct device *dev);
+#endif
+};
+
+struct bcm2711_uart_data {
+	DEVICE_MMIO_RAM; /* Must be first */
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t callback;
+	void *cb_data;
+#endif
+};
+
+static uint32_t uart_read32(const struct device *dev, uint32_t offs)
+{
+	return sys_read32(DEVICE_MMIO_GET(dev) + offs);
+}
```

Comment:
please don't add useless helpers around. Add a new variable in the `data` struct and save the `DEVICE_MMIO_GET(dev)` address in there.

Reply:
I added a function wrapper that reduces the number of calls to `DEVICE_MMIO_GET`. Furthermore, I unified the arguments of some functions to be "dev", which reduced the need for function wrapping, such as `.poll_out` and `.irq_tx_ready`. So This function is not useless.  If a new variable is added to the structure, every time the variable is needed, it would require extracting the variable from the structure and an additional type conversion.

Reply:
I added a function wrapper that reduces the number of calls to DEVICE_MMIO_GET. Furthermore, I unified the arguments of some functions to be "dev", which reduced the need for function wrapping, such as .poll_out and .irq_tx_ready. So, This function is not unuseless.  If a new variable is added to the structure, every time the variable is needed, it would require extracting the variable from the structure and an additional type conversion.
```c

struct bcm2711_uart_data {
	DEVICE_MMIO_RAM; /* Must be first */
        mem_addr_t base;
#ifdef CONFIG_UART_INTERRUPT_DRIVEN
	uart_irq_callback_user_data_t callback;
	void *cb_data;
#endif
};

static int uart_bcm2711_init(const struct device *dev)
{
	const struct bcm2711_uart_config *uart_cfg = (dev)->config;
	struct bcm2711_uart_data *data = dev->data;

	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
	data->base = DEVICE_MMIO_GET(dev);
	bcm2711_mu_lowlevel_init(dev, 1, uart_cfg->baud_rate, uart_cfg->clocks);
	return 0;
}
```

every time i wanna use `base`, i must 
```c
static void uart_bcm2711_poll_out(const struct device *dev, unsigned char c)
{
	struct bcm2711_uart_data *data = dev->data;
	bcm2711_mu_lowlevel_putc(data->base, c);
}
```
There will be an additional local variable, but its significance is minimal.

Reply:
the whole point of the `data` structure for the device driver is exactly that: you pass around only the `struct device` and then you can get all the context out by using the associated `data` structure, now that could be nitpicking or bike shedding for sure but you are here adding two more functions not even inlined to basically wrap around IO helpers. Try to compile that and look at the generated assembly code to see how wasteful is that.

Now, you don't want to use the `data` structure for religious reasons? fine to me. But at least make those inlined functions.

Reply:
No more reasons, Just don't understand why add a new variable. OK, got it now, will update later .